src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx

```diff
@@ -2696,7 +2743,118 @@ void HwEmShim::closemMessengerThread() {
+    if(buf != NULL)
+        free(buf);
+    free(i_buf);
+    free(ri_buf);
+    delete ci_msg;
+    delete ri_msg;
+    delete Q2h_sock;
+}
+int Q2H_helper::poolingon_Qdma() {
+    //Pooling on socket for any memory or interrupt requests from SIM_QDMA
+    memset(i_buf,0,i_len);
+    r = Q2h_sock->sk_read(i_buf, i_len);
+    if (r <= 0) {
+    	return r;
+    }
+    assert(i_len == (uint32_t)r);
+    ci_msg->ParseFromArray(i_buf, i_len);
+    if (ci_msg->xcl_api() == xclClose_n) {
+        return -1;
+    }
+    if (buf_size == 0) {
+    	buf = malloc(ci_msg->size());
+        buf_size = ci_msg->size();
+    } else if (buf_size < ci_msg->size()) {
+    	buf = (void*) realloc(buf, ci_msg->size());
+        buf_size = ci_msg->size();
+    }
+    memset(buf,0,buf_size);
+    r = Q2h_sock->sk_read(buf, ci_msg->size());
+    assert((uint32_t)r == ci_msg->size());
+    select_msg();
+    return 1;
+}
+void Q2H_helper::select_msg() {
+    if (ci_msg->xcl_api() == xclQdma2HostReadMem_n) {
+    	xclSlaveReadReq_call c_msg;
+    	c_msg.ParseFromArray(buf, r);
+        void *data = malloc(c_msg.size());
+        bool resp = (*m_q2h_slavetransaction)(false, 0,(unsigned long int)c_msg.addr(),data,(unsigned long int)c_msg.size());
+        xclSlaveReadReq_response r_msg;
+        r_msg.set_valid(resp);
+        r_msg.set_data(data,c_msg.size());
+        int r_len = r_msg.ByteSize();
+        SEND_RESP2QDMA();
+        free(data);
+    }        
+    if (ci_msg->xcl_api() == xclQdma2HostWriteMem_n) {
+    	xclSlaveWriteReq_call c_msg;
+    	c_msg.ParseFromArray(buf, r);
+        bool resp = (*m_q2h_slavetransaction)(false, 1,(unsigned long int)c_msg.addr(),(void*)c_msg.data().c_str(),(unsigned long int)c_msg.size());
+        xclSlaveWriteReq_response r_msg;
+        r_msg.set_valid(resp);
+        int r_len = r_msg.ByteSize();
+        SEND_RESP2QDMA();
+    }        
+    if (ci_msg->xcl_api() == xclQdma2HostInterrupt_n) {
+    	xclInterruptOccured_call c_msg;
+    	c_msg.ParseFromArray(buf, r);
+        void *interrupt_line_ptr = malloc(4);
+        *((uint32_t*)interrupt_line_ptr) = c_msg.interrupt_line();
+        bool resp = (*m_q2h_slavetransaction)(true, 1,0,interrupt_line_ptr,4);
+        free(interrupt_line_ptr);
+        xclInterruptOccured_response r_msg;
+        r_msg.set_valid(resp);
+        int r_len = r_msg.ByteSize();
+        SEND_RESP2QDMA();
+    }        
+}
+void Q2H_helper::register_q2h_slavetransaction(bool (*q2h_slavetransaction)(bool, int, unsigned long int, void*,unsigned long int)) {
+    m_q2h_slavetransaction = q2h_slavetransaction;
+}
+
+void hostMemAccessThread(xclhwemhal2::HwEmShim* inst) {
+    Q2H_helper* mq2h_helper_ptr = new Q2H_helper;
+    mq2h_helper_ptr->register_q2h_slavetransaction(device2xrt_trans_cb);
+    while(1){
+        if (!inst->get_simulator_started())
+            return;
+        int r = mq2h_helper_ptr->poolingon_Qdma();
+        if(r < 0)
+            break;
+    }
+    delete mq2h_helper_ptr;
```

Comment:
Not exception safe

Reply:
Now its under try catch block to catch exceptions.